### PR TITLE
add vault tests

### PR DESCRIPTION
**Pull request by ttarsi**
Core is straightforward, we have `CoreFixture` and a `CoreTest`
For Vaults, we have a `BasicVaultTest` contract which doesn't deploy a vault, but defines the functions any vault should pass. It is an abstract contract that can be inherited – children can deploy their own vaults with any configuration and run extra tests based on that config.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,9 @@
+test: build
```
Comment (ttarsi) [inline Makefile:1]: seems like there's a bit of a forge bug where the vault tests hang (maybe a result of them all inheriting the same abstract contract ?), but this runs each of them sequentially

File: src/test/Core.t.sol [hunk 2]
```diff
@@ -208,28 +215,12 @@ contract CoreTest is CoreFixture {
 
         core.grantRole(C.WHITELISTED_ROLE, user);
         assertTrue(core.isWhitelisted(user));
+
+        vm.stopPrank();
```
Comment (kevinhalliday) [inline src/test/Core.t.sol:219]: it wouldn't add much value in this case, but - would be nice if function modifiers could enforce post-requisites (is that a word?) and not just pre-requisite. I want for more python-decorator-esque behavior

File: src/test/utils/VaultUtils.sol [hunk 3]
```diff
@@ -1,152 +1,453 @@
...
+        vault.depositToken0(0);
+    }
 
-    function setUp() public virtual override {
-        super.setUp();
+    function test_depositToken0WithValueRevert() public {
+        if (!vault.isNativeVault()) {
+            vm.expectRevert("NOT_NATIVE_VAULT");
+            vault.depositToken0{ value: 1 }(0);
+        }
```
Comment (kevinhalliday) [inline src/test/utils/VaultUtils.sol:68]: what is the expected behavior for `depositToken0(<some non zero value>)` on a native vault

File: src/test/utils/VaultUtils.sol [hunk 4]
```diff
@@ -1,152 +1,453 @@
...
+
+        simulateFees(amount, amount);
+        withdrawToken0();
+        withdrawToken1();
+
+        advance();
+
+        (uint256 token0Fees, uint256 token1Fees) = vault.feesAccrued();
+        assertTrue(token0Fees > 0);
+        assertEq(token1Fees, 0);
```
Comment (kevinhalliday) [inline src/test/utils/VaultUtils.sol:249]: Is this always true? or is it dependent on `CEIL` & `FLOOR` configurations?

File: src/test/utils/VaultUtils.sol [hunk 5]
```diff
@@ -1,152 +1,453 @@
...
-                            0
-                        )
-                    )
-                )
-            )
-        );
+    function simulateFees(uint256 token0Amount, uint256 token1Amount) internal {
+        getToken0(address(pair), token0Amount);
+        getToken1(address(pair), token1Amount);
+        pair.sync();
```
Comment (kevinhalliday) [inline src/test/utils/VaultUtils.sol:451]: Great file. Really great file

**Review by kevinhalliday — APPROVED** [review]
Tests are comprehensive and easy to understand - nice work. I'm liking this forge thing. Left a couple comments, nothing major. 

One workflow note - would've been a bit easier to determine which diffs needed attention and which were just reformatting had the `printWidth` update been in a separate PR. But who am i to complain - I'm the king cumbersome PRs.

File: src/test/utils/VaultUtils.sol (hunk 3 above)
Comment (ttarsi) [reply src/test/utils/VaultUtils.sol:68]: fair point, it ignores that value and is not expected to revert. I'll rename to be more accurate

File: src/test/Core.t.sol (hunk 2 above)
Comment (ttarsi) [reply src/test/Core.t.sol:219]: ooh yes we can do that
```
modifier prankAs(address user) {
  vm.startPrank(user);
  _;
  vm.stopPrank();
}
```

then we can just create a function like:
function xyz() prankAs(governor) { abc };

**Review by kevinhalliday — APPROVED** [review]
👌

**Pull request merged**